inc/XamlApplication.cpp

```diff
@@ -0,0 +1,165 @@
+  XamlApplication::XamlApplication(winrt::Windows::Foundation::Collections::IVector<winrt::Windows::UI::Xaml::Markup::IXamlMetadataProvider> providers) {
+    for (auto provider : providers) {
+      m_providers.Append(provider);
+    }
+
+    Initialize();
+  }
+
+  XamlApplication::XamlApplication() {}
+
+  void XamlApplication::Initialize() {
+    const auto out = outer();
+    if (out) {
+      winrt::Windows::UI::Xaml::Markup::IXamlMetadataProvider provider(nullptr);
+      winrt::check_hresult(out->QueryInterface(
+        winrt::guid_of<winrt::Windows::UI::Xaml::Markup::IXamlMetadataProvider>(),
+        reinterpret_cast<void**>(winrt::put_abi(provider))));
+
+      winrt::Windows::UI::Xaml::Markup::IXamlMetadataProvider selfProvider(nullptr);
+      winrt::check_hresult(this->NonDelegatingQueryInterface(
+        winrt::guid_of<winrt::Windows::UI::Xaml::Markup::IXamlMetadataProvider>(),
+        reinterpret_cast<void**>(winrt::put_abi(selfProvider))));
+
+      if (get_abi(selfProvider) != get_abi(provider)) {
+        m_providers.Append(provider);
+      }
+    }
+
+    const auto dispatcherQueue = winrt::Windows::System::DispatcherQueue::GetForCurrentThread();
+    if (!dispatcherQueue) {
+      m_executionMode = ExecutionMode::Win32;
+      m_windowsXamlManager = ::winrt::Windows::UI::Xaml::Hosting::WindowsXamlManager::InitializeForCurrentThread();
+
+      if (auto coreWindow = winrt::Windows::UI::Core::CoreWindow::GetForCurrentThread()) {
+        auto coreWindowInterop = coreWindow.as<ICoreWindowInterop>();
+        HWND coreWindowHwnd = nullptr;
+        coreWindowInterop->get_WindowHandle(&coreWindowHwnd);
+        ::ShowWindow(coreWindowHwnd, SW_HIDE);
+      }
+    } else {
+      m_executionMode = ExecutionMode::UWP;
+    }
+  }
+
+  //void XamlApplication::Run(winrt::Microsoft::Toolkit::Win32::UI::XamlHost::DesktopWindow window) {
+  //  window.Show();
+  //  Run();
+  //}
+
+  void XamlApplication::Run() {
+    MSG msg = {};
+    while (GetMessage(&msg, nullptr, 0, 0)) {
+      TranslateMessage(&msg);
+      DispatchMessage(&msg);
+    }
+
+    return; //(int)msg.wParam;
+  }
+
+  winrt::Windows::Foundation::IClosable XamlApplication::WindowsXamlManager() const {
+    return m_windowsXamlManager;
+  }
+
+  void XamlApplication::Close() {
+    if (m_bIsClosed) {
+      return;
+    }
+
+    m_bIsClosed = true;
+
+    m_windowsXamlManager.Close();
+    m_providers.Clear();
+    m_windowsXamlManager = nullptr;
+
+    Exit();
+    {
+      MSG msg = {};
+      while (PeekMessageW(&msg, nullptr, 0, 0, PM_REMOVE)) {
+        ::DispatchMessageW(&msg);
+      }
+    }
+  }
+
+  XamlApplication::~XamlApplication() {
+    Close();
+  }
+
+  ::winrt::Windows::UI::Xaml::Markup::IXamlType XamlApplication::GetXamlType(::winrt::Windows::UI::Xaml::Interop::TypeName const& type) {
+    for (const auto& provider : m_providers) {
+      const auto xamlType = provider.GetXamlType(type);

```

Comment:
can be `if (const auto xamlType = provider.GetXamlType(type))`.

here and below.